### PR TITLE
RMT: Minor cleanups, simplify address calculations, extend HIL test

### PR DESCRIPTION
**Pull request by wisp3rwind**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [ ] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs developer guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖
Various refactorings of the RMT driver that make it more readable (in my opinion) or improve error handling in edge cases.

This touches quite a bit of code, but it should be fairly easy to review commit-by-commit. Let me know if I should split this into several PRs.

#### Description
This contains
- A few minor cleanups (mostly related to `cfg` attributes).
- Simplifications to address calculations: This should make the code much more readable.
- Fixes to a few edge cases of the code.
- Extensions to the HIL test such that they should cover more code paths of the RMT driver, in particular regarding wrapping of the buffer and the recently introduced extended channel RAM support.

For details, see the individual commit messages.

Note that I disabled one of the newly added HIL tests on `esp32`: It fails in CI, and I couldn't easily figure out why (from a quick glance at the idf code and the TRM). However, it appears that this failure is not a result of any change in this PR: In https://github.com/esp-rs/esp-hal/pull/3484, I pushed a branch which contains `main` just before merging https://github.com/esp-rs/esp-hal/pull/3453 with a single additional commit which adds the tests from this PR, but leaves the driver untouched. Thus, the failure already existed after https://github.com/esp-rs/esp-hal/pull/3292. (Since it involves using a channel with more than 1 block of the RAM, the test can't be run before that PR. The underlying issue might have been around before.)

There should be no user-visible changes.

#### Testing
Build-tested esp32c3, no tests on real hardware except for CI.

### CODE REVIEW

File: esp-hal/src/rmt.rs [hunk 1]
```diff
@@ -622,47 +631,47 @@ where
...
-            {
-                unsafe {
-                    ptr.add(idx).write_volatile(*entry);
-                }
-            }
-
-            self.index +=
-                constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize / 2;
+                .map(|(idx, entry)| unsafe { ptr.add(idx).write_volatile(*entry) })
+                .count();
```
Comment (bugadani) [inline esp-hal/src/rmt.rs:661]: This makes sense but reads funny to me. Your intention is to avoid the need for an extra `let count = self.remaining_data.len().min(memsize / 2)` but I feel like the resulting iterator is a worse fit. For one, it's slightly surprising that you use `map` to actually write to the hardware. For the other, `count` *may* be less efficient if the compiler really turns it into `loop { write(); count += 1 }`.

File: esp-hal/src/rmt.rs [hunk 2]
```diff
@@ -622,47 +631,47 @@ where
...
+
+            // If count == memsize / 2 codes were written, update ram_index as
+            // - 0 -> memsize / 2
+            // - memsize / 2 -> 0
+            // Otherwise, for count < memsize / 2, the new position is invalid but the new
+            // slice is empty and we won't use ram_index again.
+            self.ram_index = memsize / 2 - self.ram_index;
+            self.remaining_data = &self.remaining_data[count..];
+            assert!(
```
Comment (bugadani) [inline esp-hal/src/rmt.rs:670]:
```suggestion
            debug_assert!(
```

File: esp-hal/src/rmt.rs [hunk 3]
```diff
@@ -1608,32 +1610,24 @@ pub trait TxChannelInternal {
...
-            .take(constants::RMT_CHANNEL_RAM_SIZE * Self::memsize() as usize)
+            .take(memsize)
             .enumerate()
-        {
-            unsafe {
-                ptr.add(idx).write_volatile(*entry);
-            }
-        }
+            .map(|(idx, entry)| unsafe { ptr.add(idx).write_volatile(*entry) })
+            .count();
```
Comment (bugadani) [inline esp-hal/src/rmt.rs:1620]: See my other comment

File: hil-test/tests/rmt.rs [hunk 4]
```diff
@@ -6,120 +6,204 @@
...
-    gpio::Level,
-    rmt::{PulseCode, Rmt, RxChannel, RxChannelConfig, TxChannel, TxChannelConfig},
+    gpio::{Level, NoPin},
+    rmt::{Error, PulseCode, Rmt, RxChannel, RxChannelConfig, TxChannel, TxChannelConfig},
     time::Rate,
 };
 use hil_test as _;
 
+cfg_if::cfg_if! {
+    if #[cfg(feature = "esp32h2")] {
```
Comment (bugadani) [inline hil-test/tests/rmt.rs:19]: I believe we define the metadata symbols for hil tests, so you should be able to use `cfg(esp32h2)` in tests as well. You don't have to, though, this is fine as is.

**Review by bugadani — COMMENTED** [review]
Thank you, these changes look pretty good! I do have a few minor comments but nothing terribly important I think.

File: hil-test/tests/rmt.rs (hunk 4 above)
Comment (wisp3rwind) [reply hil-test/tests/rmt.rs:19]: Indeed, they are defined. If I remember correctly, this was using the `feature = ...` form before, and I just kept it. But it's an easy enough change to add here; I'll do so :+1:

File: esp-hal/src/rmt.rs (hunk 1 above)
Comment (wisp3rwind) [reply esp-hal/src/rmt.rs:661]: Yup, you got my intention right there; it also felt a little less error-prone due to not duplicating length calculation.
I agree regarding mutating in the iterator, that also feels unusual to me.

I'm pushing a modified version that is closer to the old code.

**Review by bugadani — APPROVED** [review]
Thanks!

**Comment by wisp3rwind** [comment]
Test failure is on esp32c2 only, which doesn't seem to have an RMT, seemingly due do https://github.com/esp-rs/esp-hal/pull/3500.

**Comment by bugadani** [comment]
> Test failure is on esp32c2 only, which doesn't seem to have an RMT, seemingly due do https://github.com/esp-rs/esp-hal/pull/3500.

Yeah the C2 failure looks to be caused by a silent compiler update. I guess we used to run close to the time limit of the test, and the generated code got unexpectedly slower. No matter.

**Comment by wisp3rwind** [comment]
Rebased on `main`, now `esp32` fails with
```
Error: An error with the flashing procedure has occurred.

Caused by:
    Flash content verification failed.
```

**Review by MabezDev — APPROVED** [review]
Thank you!
